File: DocumentLayoutHelper/VisualBasic/RenderedDocument.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text

Imports Aspose.Words
Imports Aspose.Words.Layout

Namespace Aspose.Words.Layout
	''' <summary>
	''' Provides an API wrapper for the LayoutEnumerator class to access the page layout entities of a document presented in
	''' a object model like design.
	''' </summary>
	Public Class RenderedDocument
		Inherits LayoutEntity
		''' <summary>
		''' Creates a new instance from the supplied Aspose.Words.Document class.
		''' </summary>
		''' <param name="document">A document whose page layout model to enumerate.</param>
		''' <remarks><para>If page layout model of the document hasn't been built the enumerator calls <see cref="Document.UpdatePageLayout"/> to build it.</para>
		''' <para>Whenever document is updated and new page layout model is created, a new enumerator must be used to access it.</para></remarks>
		Public Sub New(ByVal doc As Document)
			mEnumerator = New LayoutEnumerator(doc)
			ProcessLayoutElements(Me)
			CollectLinesAndAddToMarkers()
			LinkLinesToNodes(doc)
		End Sub

		''' <summary>
		''' Provides access to the pages of a document.
		''' </summary>
		Public ReadOnly Property Pages() As LayoutCollection(Of RenderedPage)
			Get
				Return GetChildNodes(Of RenderedPage)()
			End Get
		End Property

		''' <summary>
		''' Returns all lines of the specified document paragraph.
		''' </summary>
		''' <remarks>Note that this method sometimes won't return all lines if the paragraph is inside a table.</remarks>
		Public Function GetLinesOfParagraph(ByVal para As Paragraph) As LayoutCollection(Of RenderedLine)
			mEnumerator.MoveNode(para)
			Dim lines As List(Of RenderedLine) = New List(Of RenderedLine)()

			If mLayoutToLinesLookup.ContainsKey(mEnumerator.Current) Then
				lines = mLayoutToLinesLookup(mEnumerator.Current)
			End If

			Return New LayoutCollection(Of RenderedLine)(lines)
		End Function

		Private Sub ProcessLayoutElements(ByVal current As LayoutEntity)
			Do
				Dim child As LayoutEntity = current.AddChildEntity(mEnumerator)

				If mEnumerator.MoveFirstChild() Then
					current = child

					ProcessLayoutElements(current)
					mEnumerator.MoveParent()

					current = current.Parent
				End If
			Loop While mEnumerator.MoveNext()
		End Sub

		Private Sub CollectLinesAndAddToMarkers()
			CollectLinesOfMarkersCore(LayoutEntityType.Column)
			CollectLinesOfMarkersCore(LayoutEntityType.HeaderFooter)
			CollectLinesOfMarkersCore(LayoutEntityType.Comment)
		End Sub

		Private Sub CollectLinesOfMarkersCore(ByVal type As LayoutEntityType)
			Dim collectedLines As List(Of RenderedLine) = New List(Of RenderedLine)()

			For Each page As RenderedPage In Pages
				For Each story As LayoutEntity In page.GetChildEntities(type, False)
					For Each line As RenderedLine In story.GetChildEntities(LayoutEntityType.Line, True)
						collectedLines.Add(line)
						For Each span As RenderedSpan In line.Spans
							If span.Kind = "PARAGRAPH" OrElse span.Kind = "ROW" OrElse span.Kind = "CELL" OrElse span.Kind = "SECTION" Then
								mLayoutToLinesLookup.Add(span.LayoutObject, collectedLines)
								collectedLines = New List(Of RenderedLine)()
							End If
						Next span
					Next line
				Next story
			Next page
		End Sub

		Private Sub LinkLinesToNodes(ByVal doc As Document)
			For Each para As Paragraph In doc.GetChildNodes(NodeType.Paragraph, True)
				mEnumerator.MoveNode(para)

				If mLayoutToLinesLookup.ContainsKey(mEnumerator.Current) Then
					For Each entity As LayoutEntity In mLayoutToLinesLookup(mEnumerator.Current)
						CType(entity, NodeReferenceLayoutEntity).Paragraph = para
					Next entity
				End If
			Next para
		End Sub

		Private mEnumerator As LayoutEnumerator
		Private Shared mLayoutToLinesLookup As Dictionary(Of Object, List(Of RenderedLine)) = New Dictionary(Of Object, List(Of RenderedLine))()
	End Class
End Namespace

File: DocumentLayoutHelper/CSharp/RenderedDocument.cs
using System;
using System.Collections.Generic;
using System.Text;

using Aspose.Words;
using Aspose.Words.Layout;

namespace Aspose.Words.Layout
{
    /// <summary>
    /// Provides an API wrapper for the LayoutEnumerator class to access the page layout entities of a document presented in
    /// a object model like design.
    /// </summary>
    public class RenderedDocument : LayoutEntity
    {
        /// <summary>
        /// Creates a new instance from the supplied Aspose.Words.Document class.
        /// </summary>
        /// <param name="document">A document whose page layout model to enumerate.</param>
        /// <remarks><para>If page layout model of the document hasn't been built the enumerator calls <see cref="Document.UpdatePageLayout"/> to build it.</para>
        /// <para>Whenever document is updated and new page layout model is created, a new enumerator must be used to access it.</para></remarks>
        public RenderedDocument(Document doc)
        {
            mEnumerator = new LayoutEnumerator(doc);
            ProcessLayoutElements(this);
            CollectLinesAndAddToMarkers();
            LinkLinesToNodes(doc);
        }

        /// <summary>
        /// Provides access to the pages of a document.
        /// </summary>
        public LayoutCollection<RenderedPage> Pages
        {
            get
            {
                return GetChildNodes<RenderedPage>();
            }
        }

        /// <summary>
        /// Returns all lines of the specified document paragraph.
        /// </summary>
        /// <remarks>Note that this method sometimes won't return all lines if the paragraph is inside a table.</remarks>
        public LayoutCollection<RenderedLine> GetLinesOfParagraph(Paragraph para)
        {
            mEnumerator.MoveNode(para);
            List<RenderedLine> lines = new List<RenderedLine>();

            if (mLayoutToLinesLookup.ContainsKey(mEnumerator.Current))
                lines = mLayoutToLinesLookup[mEnumerator.Current];

            return new LayoutCollection<RenderedLine>(lines);
        }

        private void ProcessLayoutElements(LayoutEntity current)
        {
            do
            {
                LayoutEntity child = current.AddChildEntity(mEnumerator);

                if (mEnumerator.MoveFirstChild())
                {
                    current = child;

                    ProcessLayoutElements(current);
                    mEnumerator.MoveParent();

                    current = current.Parent;
                }
            } while (mEnumerator.MoveNext());
        }

        private void CollectLinesAndAddToMarkers()
        {
            CollectLinesOfMarkersCore(LayoutEntityType.Column);
            CollectLinesOfMarkersCore(LayoutEntityType.HeaderFooter);
            CollectLinesOfMarkersCore(LayoutEntityType.Comment);
        }

        private void CollectLinesOfMarkersCore(LayoutEntityType type)
        {
            List<RenderedLine> collectedLines = new List<RenderedLine>();

            foreach (RenderedPage page in Pages)
            {
                foreach (LayoutEntity story in page.GetChildEntities(type, false))
                {
                    foreach (RenderedLine line in story.GetChildEntities(LayoutEntityType.Line, true))
                    {
                        collectedLines.Add(line);
                        foreach (RenderedSpan span in line.Spans)
                        {
                            if (span.Kind == "PARAGRAPH" || span.Kind == "ROW" || span.Kind == "CELL" || span.Kind == "SECTION")
                            {
                                mLayoutToLinesLookup.Add(span.LayoutObject, collectedLines);
                                collectedLines = new List<RenderedLine>();
                            }
                        }
                    }
                }
            }
        }

        private void LinkLinesToNodes(Document doc)
        {
            foreach (Paragraph para in doc.GetChildNodes(NodeType.Paragraph, true))
            {
                mEnumerator.MoveNode(para);

                if (mLayoutToLinesLookup.ContainsKey(mEnumerator.Current))
                {
                    foreach (LayoutEntity entity in mLayoutToLinesLookup[mEnumerator.Current])
                    {
                        ((NodeReferenceLayoutEntity)entity).Paragraph = para;
                    }
                }
            }
        }

        private LayoutEnumerator mEnumerator;
        private static Dictionary<object, List<RenderedLine>> mLayoutToLinesLookup =
                new Dictionary<object, List<RenderedLine>>();
    }
}
